### PR TITLE
mobile/ci: Improve mobile jobs

### LINKED ISSUE #24437: Dont run (mobile) github actions (for required checks) if there is no need
In #24407 a change was made to not use the `paths` feature in github actions to exclude checks

this was done to ensure that the checks could be `required` for PRs to pass

the problem with this approach is that it pulls a full copy of the repo and is much slower than using githubs paths (which afaict uses its own diff algorithm against the target branch, and perhaps pre-checksout the repo - not sure about this exactly)

in more recent version of the github `actions/checkout` action the default is to pull the repo with a `depth` of 1, so the current `mobile/should_run_ci.sh` script, where it assumes that the full repo is checked out, breaks (this is easy enough to workaround but just hightlights the issue)

i think we can speed this up and reduce the load on CI considerably by reverting to using the `paths` selector and adding the github way of allowing a required check to pass when skipped (https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/defining-the-mergeability-of-pull-requests/troubleshooting-required-status-checks#handling-skipped-but-required-checks)

like the azp way its a bit kludgy but it think it would be quite a bit faster/more efficient

### PR DESCRIPTION
**Pull request by phlax**
Fix #24437

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by phlax** [comment]
this will need some required job name changes before it can land

**Comment by phlax** [comment]
this has a few collateral benefits:

- speeds up all ci as it doesnt have to repeatedly checkout the whole envoy repo
- particularly speeds up noop ci as it currently checks out the whole repo to decide to noop
- improves ordering in the PR checks box
- provides a much better github ci ui for tracking a commit or PR as it goes through CI
- allows all jobs for a pr/commit to be started/stopped/restarted/retested in on place

the main downside is that you cant stop an individual job - only the entire workflow - but this is mostly a downside when compared to azp

**Comment by phlax** [comment]
image

**Comment by phlax** [comment]
image

**Comment by alyssawilk** [comment]
so I like the speed up CI by not doing repeated check outs, but as someone doing mobile development not being able to restart individual jobs is a huge hit and IMO not worth the speed up (especially given how much faster mobile CI is).  cc @RyanTheOptimist @abeyad for thoughts.

**Comment by phlax** [comment]
> IMO not worth the speed up

tbh the speedup is a secondary consideration - its the unmanageability of it that has motivated be to do this

currently you have to click so many places trying to find what has/not failed - and as someone who works on envoy main ci, i desparately need a way to cancel mobile ci

**Comment by phlax** [comment]
note - nicely orderered ci jobs

image

**Comment by phlax** [comment]
ah, the other consideration

atm mobile ci flakes a huge amount - this is also a step to caching as we have for envoy to pretty good effect

**Comment by alyssawilk** [comment]
Sorry I don't quite understand the problem with failure, asking some clarifying questions.  
When I want to know what failed, I scroll down the list to where there's an x.  Is your workflow problem that you have to click once per failed job?  I'm also not sure why you'd need to cancel mobile ci, but if having to click per job is the problem could we add a cancel command similar to the retest command, and then not lose retryability for mobile devs?

**Comment by alyssawilk** [comment]
caching would be great but isn't that orthogonal from consolidation?

**Comment by phlax** [comment]
> When I want to know what failed, I scroll down the list to where there's an x.

which tends to  be jumbled up - has different icons that are not always clear - ie skipped/timedout etc - this lets you see the mobile ci in one place - and track its progress - atm the jobs _are_ dependent but there is no sense of that

the other problem is if you click on the actions for a pr - you get a huge list of mobile actions - this resolves that
 and lets you see the mobile jobs separately - which is gonna make it a lot easier to move envoy jobs

**Comment by phlax** [comment]
> caching would be great but isn't that orthogonal from consolidation?

not really because you need a pre-jobs as done here to prime the cache/s (at least for the build image)

**Comment by phlax** [comment]
> ut if having to click per job is the problem could we add a cancel command similar to the retest command, and then not lose retryability for mobile devs?

why not just leverage the github ui, rather than working against it - the `commands` job already has issues and needswork

**Comment by phlax** [comment]
the point is **this is a workflow, not 15**

**Comment by phlax** [comment]
if it helps at all - there is no dependency here (other than the init job) that is not in existing mobile workflows

**Comment by RyanTheOptimist** [comment]
> so I like the speed up CI by not doing repeated check outs, but as someone doing mobile development not being able to restart individual jobs is a huge hit and IMO not worth the speed up (especially given how much faster mobile CI is). cc @RyanTheOptimist @abeyad for thoughts.

Losing the ability to restart individual job would be a REAL bummer. I notice regularly that when a mobile job fails, I can restart it and often it finishes before the first run of some longer job. So the net time to pass CI is unchanged. However, when I have a PR with a non-mobile failure, I have to wait until everything finishes (which can often be very time consuming) before I'm able to restart. This is a major bummer.

I'd love to get a cleaner UI, which this PR gives. But it doesn't feel worth the productivity hit. Is there any middle ground we could pursue somehow?

**Comment by phlax** [comment]
putting aside the too-many-workflows issue we can get ~all of the performance benefit from here by using a `workflow_run` pattern

its quite a bit more complex, but at least it addresses the repeatedly checkout the repo issue - and gives a modicum of control - you get ~30s to cancel the workflow before it triggers all the others

this will keep the workflows separate - we can cross that bridge after

**Comment by phlax** [comment]
> using a workflow_run pattern

this is not an ideal solution as it has similar problems to pull_request_target - it runs privileged by default

the other problem is that the triggered workflows must be on `main` which means that we cant maintain separate ci for each release branch (at least not easily), and it makes testing much harder

**Comment by phlax** [comment]
experimenting and checking docs further there is no easy/safe way to trigger ci using any of the other gh events - only workflow_dispatch/repository_dispatch create new workflows, only workflow_call retains the correct context

workflow_dispatch could kinda be made to work, but only being served from `main`

tldr these are only really useful for publishing jobs, as it would otherwise lock all release branches to using the ci from `main` (at least to initialize)

**Comment by phlax** [comment]
plan c - which is less performant, but still saves a lot of resources compared to now is to just check for changes once per workflow - not per job as is done now

**Comment by phlax** [comment]
working on this - the problem is actually a lot worse than i understood - in most/many of the jobs its running the should_run script _inside_ the mobile/build container

**Comment by phlax** [comment]
i ran a test pr #27449 to ensure that the required jobs pass when skipped

**Comment by phlax** [comment]
/retest

```console
//test/swift/integration:reset_connectivity_state_test                   FAILED in 80.4s
```

https://github.com/envoyproxy/envoy/actions/runs/5002203521/jobs/8961876092?pr=27404#step:4:4136

```console
Error: Final attempt failed. Timeout of 600000ms hit
```

https://github.com/envoyproxy/envoy/actions/runs/5002203534/jobs/8961981242?pr=27404#step:5:6568

**Comment by phlax** [comment]
```console
 ERROR: The Build Event Protocol upload failed: All 4 retry attempts failed. DEADLINE_EXCEEDED: DEADLINE_EXCEEDED: deadline exceeded after 14.999889499s. [closed=[], open=[[buffered_nanos=47600, remote_addr=envoy.cluster.engflow.com/54.197.155.115:443]]] DEADLINE_EXCEEDED: DEADLINE_EXCEEDED: deadline exceeded after 14.999889499s. [closed=[], open=[[buffered_nanos=47600, remote_addr=envoy.cluster.engflow.com/54.197.155.115:443]]]
```

**Comment by phlax** [comment]
> /wait unmerged 

ill merge now ...

> ... and unsure of status

keen to land 8)

**Comment by alyssawilk** [comment]
this still merges all the jobs so we can't retry tasks individually while other tasks are running?

**Comment by phlax** [comment]
> this still merges all the jobs so we can't retry tasks individually while other tasks are running?

no - it runs each as a separate workflow - the difference is that it only checks what to run once per workflow

File: mobile/tools/what_to_run.sh [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        echo "mobile_coverage=true"
+        echo "mobile_formatting=true"
+        echo "mobile_ios_build=true"
+        echo "mobile_ios_tests=true"
+        echo "mobile_release_validation=true"
+        echo "mobile_tsan=true"
+    } >> "$GITHUB_OUTPUT"
+}
+
+run_ci_for_changed_files () {
```
Comment (alyssawilk) [inline mobile/tools/what_to_run.sh:27]: can we add some breadcrumb comments while we're in here?  We only run the expensive CI jobs for "high risk" changes: mobile PRs or PRs which touch the files listed in CHANGE_MATCH

File: mobile/tools/what_to_run.sh [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    run_default_ci
+    {
+        echo "mobile_android_build_all=true"
+        echo "mobile_ios_build_all=true"
+    } >> "$GITHUB_OUTPUT"
+}
+
+if [[ "$BRANCH_NAME" == "main" ]]; then
```
Comment (alyssawilk) [inline mobile/tools/what_to_run.sh:35]: comment about not running for release branches?

File: mobile/tools/what_to_run.sh [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+BRANCH_NAME="$GITHUB_REF_NAME"
+BASE_COMMIT="$(git merge-base origin/main HEAD)"
+CHANGED_FILES="$(git diff "${BASE_COMMIT}" --name-only)"
+CHANGE_MATCH='^mobile/|^bazel/repository_locations\.bzl|^\.bazelrc|^\.github/workflows/mobile-*|^\.github/workflows/env.yml'
+
+
+run_default_ci () {
+    {
+        echo "mobile_android_build=true"
```
Comment (alyssawilk) [inline mobile/tools/what_to_run.sh:13]: do we still skip these on docs only PRs etc?

Comment (phlax) [reply mobile/tools/what_to_run.sh:13]: i believe the conditions are the same - /docs is not in the `CHANGE_MATCH` so it would be skipped

File: mobile/tools/what_to_run.sh (hunk 2 above)
Comment (phlax) [reply mobile/tools/what_to_run.sh:45]: i added a comment about release branches in the explainer

**Comment by phlax** [comment]
/retest mobile

for the same failure again

**Review by alyssawilk — APPROVED** [review]
